### PR TITLE
Use existing project as a template to create a new one

### PR DESCRIPTION
**Pull request by carpodaster**
**What?**
Allows people who have submitted a project in a past season to use that project as a base for a new submission in the current year.

screen shot 2018-01-19 at 22 34 55

The action will trigger the normal "new" form with the fields being already populated:

screen shot 2018-01-19 at 22 41 25

It only works for past years. You cannot clone an already submitted project for the current season.

This said, I'm more than happy for suggestions on where to put the action button. Especially because my `colspan` hack link looks really bad when you don't have anything to use as a template:

screen shot 2018-01-19 at 22 35 09

### CODE REVIEW

**Comment by klappradla** [comment]
Haven't looked at the code yet, but concerning your problem with empty table cells: there's actually a CSS property for this - [`empty-cells`](https://css-tricks.com/almanac/properties/e/empty-cells/) 😉

File: app/helpers/projects_helper.rb [hunk 1]
```diff
@@ -21,7 +21,7 @@ def project_tags(project)
 
   # @return [Array<String>] a list of years we have projects for, most recent first.
   def project_years
-    (Season.joins(:projects).distinct.pluck(:name) + [Date.today.year.to_s]).uniq.reverse
+    (Season.joins(:projects).distinct.pluck(:name) + [Date.today.year.to_s]).uniq.sort.reverse
```
Comment (klappradla) [inline app/helpers/projects_helper.rb]: 👍

File: spec/requests/projects_spec.rb [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        expect(flash[:alert]).to match %r{not authorized}
+      end
+    end
+
+    context 'when the user is not logged in' do
+      it_behaves_like 'returns to the root page'
+    end
+
+    context 'when the user is logged in' do
+      let!(:user) { create :user }
```
Comment (klappradla) [inline spec/requests/projects_spec.rb]: No need for the bang here, the `before` hook will make sure the user is created.

File: spec/requests/projects_spec.rb [hunk 3]
```diff
@@ -0,0 +1,53 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe ProjectsController, type: :request do
+
+  describe 'GET /projects/:id/use_as_template' do
+    let!(:submitter) { create :user }
+    let!(:season)    { nil }
+    let!(:project)   { create :project, season: season, submitter: submitter }
```
Comment (klappradla) [inline spec/requests/projects_spec.rb]: If you already use `!` for the project, there's no need for *banging*  the submitter and the season.

File: spec/requests/projects_spec.rb [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    context 'when the user is logged in' do
+      let!(:user) { create :user }
+      before { sign_in user }
+
+      context 'when the user does not own the project' do
+        it_behaves_like 'returns to the root page'
+      end
+
+      context 'when the user has submitted the project before' do
+        let!(:submitter) { user }
```
Comment (klappradla) [inline spec/requests/projects_spec.rb]: As I said ☝️ I think the `!` is obsolete here.

File: spec/requests/projects_spec.rb [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+        context 'when the project is from the current season' do
+          let!(:season) { Season.current }
```
Comment (klappradla) [inline spec/requests/projects_spec.rb]: Again, as said before, I think the `!` is obsolete here.

File: spec/requests/projects_spec.rb [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+          it 'returns to the root page' do
+            get use_as_template_project_path(project)
+            expect(response).to be_success
+            expect(response.body).to match project.name
+          end
+        end
+      end
+    end
+  end
+
```
Comment (klappradla) [inline spec/requests/projects_spec.rb]: Lonely blank line 👾

File: spec/models/ability_spec.rb [hunk 7]
```diff
@@ -382,6 +382,24 @@
...
+            it { is_expected.not_to be_able_to :use_as_template, project }
+          end
+        end
+
+        context 'for a project submitted by someone else' do
+          let(:project) { build :project }
+          it { is_expected.not_to be_able_to :use_as_template, project }
+        end
+      end
     end
```
Comment (klappradla) [inline spec/models/ability_spec.rb]: Lonely blank line 👾

File: app/models/ability.rb [hunk 8]
```diff
@@ -76,6 +76,9 @@ def initialize(user)
       user.admin? ||
         (user.confirmed? && user == project.submitter)
     end
+    can :use_as_template, Project do |project|
+      user == project.submitter && !project.season&.current?
+    end
```
Comment (F3PiX) [inline app/models/ability.rb]: Would you mind adding #projects as a heading before line 75?

**Comment by carpodaster** [comment]
Thanks for your feedback, I think I tended to everything. Also thanks for introducing me to `empty-cells`, Max. I couldn't get it to work and already blamed my lame CSS skills when I learned that apparently Bootstrap's border-collapse conflicts with empty-cells. Hm.

**Pull request merged**